Content.Server/_FarHorizons/Medical/Surgery/SurgeryOverhaulSystems.cs

```diff
@@ -0,0 +1,185 @@
+        if (!surgProto.TryGetComponent<NecrosisSurgeryComponent>(out var surgProtoComp, _componentFactory)) return;
+        surgComp.RequiredSurgeries.Clear();
+
+        for (int i = 0; i < surgProtoComp.AmountOfSurgeries; i++)
+        {
+            EntProtoId chosenSurgery;
+            if (_surgeriesForRotten.Count == 0)
+                break;
+
+            while (true)
+            {
+                chosenSurgery = _random.PickAndTake(_surgeriesForRotten);
+                if (!_entity.TryGetSingleton(chosenSurgery, out var surgeryEnt))
+                    continue;
+                var ev = new SurgeryValidEvent(args.Body, args.Part);
+
+                RaiseLocalEvent(surgeryEnt, ref ev);
+
+                if (!ev.Cancelled)
+                    break;
+            }
+            surgComp.RequiredSurgeries.Add(chosenSurgery!);
+        }
+    }
+    private void OnRepairEyesComplete(EntityUid uid, SurgeryRepairEyesComponent comp, ref SurgeryStepCompleteEvent args)
+    {
+        if (TryComp<BlindableComponent>(args.Body, out var blindComp))
+            _blindableSystem.AdjustEyeDamage(args.Body, -blindComp.EyeDamage);
+    }
+    // Valid Event Checks
+    private void OnNecrosisSurgeryValid(Entity<NecrosisSurgeryComponent> ent, ref SurgeryValidEvent args)
+    {
+        if (!_rottingSystem.IsRotten(args.Body))
+        {
+            args.Cancelled = true;
+            return;
+        }
+        if (HasComp<DisableSurgeryComponent>(ent)
+            && TryComp<MetaDataComponent>(ent, out var metaComp)
+            && metaComp.EntityPrototype != null
+            && !ent.Comp.RequiredSurgeries.Contains(metaComp.EntityPrototype.ID))
+        {
+            args.Cancelled = true;
+            return;
+        }
+    }
+    private void OnResearchSurgeryValid(Entity<SurgeryTechnologyComponent> ent, ref SurgeryValidEvent args)
+    {
+        if (ent.Comp.RequiredTechnology != null)
+        {
+            var TechProto = _prototypes.Index<TechnologyPrototype>(ent.Comp.RequiredTechnology);
+            if (!TryComp(args.Body, out BuckleComponent? buckle) || !TryComp(buckle.BuckledTo, out DeviceLinkSinkComponent? linkComp) || linkComp.LinkedSources.Count == 0)
+            {
+                args.Cancelled = true;
+                return;
+            }
+            if (TryComp(linkComp.LinkedSources.First(), out TechnologyDatabaseComponent? techComp) && !_research.IsTechnologyUnlocked(args.Body, TechProto, techComp))
+            {
+                args.Cancelled = true;
+                return;
+            }
+        }
+    }
+    private void LoadSurgeriesForRotten()
+    {
+        _surgeriesForRotten.Clear();
+
+        foreach (var entity in _prototypes.EnumeratePrototypes<EntityPrototype>())
+        {
+            var surgProto = _prototypes.Index<EntityPrototype>(entity);
+            if (surgProto.HasComponent<NecrosisSurgeryStepComponent>())
+                _surgeriesForRotten.Add(new EntProtoId(entity.ID));
+        }
```

Comment:
```suggestion
        foreach (var entity in _prototypes.EnumeratePrototypes<EntityPrototype>())
            if (entity.HasComponent<NecrosisSurgeryStepComponent>())
                _surgeriesForRotten.Add(new EntProtoId(entity.ID));
```

surgProto is the same thing as entity you get in iteration, it comes from the same collection by indexing itself. This can be simplified

Reply:
Added thanks for telling.